### ISSUE TITLE
[BUG] Trakt In Progress Shows not reflecting latest updates

### ISSUE THREAD

**Issue by HairyHook** [jurialmunkey/plugin.video.themoviedb.helper#1002]
### Describe the bug

Hi Jurial,

Using the addon in both PC and Nvidia Shield. When I look into my In Progress TV Shows inside TMDbhelper, it doesn't show the latest. I can see the latest in both Trakt's website and other addons such as Fen or Ezra. Is there any way to refresh the Trakt data? I tried deleting the Trakt cache but it made no difference.

Thanks in advance. 

### Steps To Reproduce

Addon menu
1- TV Shows
2 - In Progress Shows

### Checklist

- [X] I have searched the issues and this bug has not been reported.
- [X] I have updated TMDbHelper to the most recent version.
- [X] I have described the bug and included steps to reproduce the issue
- [ ] I have uploaded my debug log.

**Comment by jurialmunkey** [comment]
What is meant by "it doesn't show the latest"?

Latest in terms of what? Air date? Watched history? Newly started shows?

**Comment by HairyHook** [comment]
Air date. For example, weekly shows such as The Last of Us. I check Trakt's website and it's listed, I check Fen's Trakt In Progress section and it is listed. TMDb Helper sometimes takes 24 to 48 hours to get them populated in my in progress widget.

**Comment by jurialmunkey** [comment]
Do you have the option enabled to use the calendar for next aired?

**Comment by HairyHook** [comment]
Appreciate you taking the time to look into this. "Next aired library calendar to local tvshows" is enabled under Settings - Widgets.

**Comment by jurialmunkey** [comment]
Is it only the "Your In-Progress Shows" or is it also the "Your Next Episodes" and "Still to Watch" lists? And are you saying it is the entire tvshow not displaying in the list, or that the newly aired episode is missing when you open the tvshow?

Also are your region settings correct in both Kodi and TMDbHelper?

One thing to try when it next occurs is to mark an episode as watched on Trakt and then restart Kodi to see if that has any impact. The refresh for some of the sync caches are tied to Trakt's last activity timestamps for watched activity (i.e. cache only updates if Trakt reports a timestamp for new activity after the timestamp for the cached sync).

I thought I addressed all the cases where that occurs but possible I missed something somewhere and the cache is not updating because you haven't watched something (so instead you're just waiting on the cache to expire).

**Comment by HairyHook** [comment]
Thank you for the details explanation! It was affecting all lists, including "Your In-Progress Shows", "Your Next Episodes" and "Still to Watch". Missing just the newly added episode. 

At the time of opening the bug, even marking something as watched would not trigger an update.

It's possible it has been fixed in a recent update. When I brought it up we were watching a show that airs weekly and TMDbhelper was the only addon missing the last episode. Within 24/48 hours, it appeared by itself. 

Patience is a virtue, if I need to wait 24h for an episode to appear, it's not a problem. I just wanted to bring it up to your attention. Ill keep an eye on the issue (I think Ted Lasso airs weekly) to see if it is still happening.

**Comment by florence-wolfe** [comment]
I would like to jump in on this issue as I've begun to notice a similar behavior and may have come across some kind of edge-case.

(based on this info please let me know if you'd like this split into its own issue)

> Kodi 20.2
> TMDBHelper v5.2.8~nexus
> Arctic Fuse v0.4.10

I frequently use the `In Progress Shows` list from TMDBHelper's Trakt integration as a widget in all of my skins. Recently, I noticed that the episode counts are not updating correctly to include the latest episode on _some_ series.

For example, anecdotally, the series `The Golden Bachelor` is showing me as fully watched with the ✅ icon despite new episodes having been released.

For context, today is Dec 1st and the new episode was released on Nov 30th.

If I select that item to view the seasons from the `Combined Landscape` view, I get the following:

- the `.. Folder` is selected by default which triggers a 404 to TMDB since it's request a `-1` season #
  - (technically this is unrelated but may be worth bringing up since it could lead to a small optimization)
- Season 1 is displayed as `2023 — Season 1 \n 8 of 8 watched`.
  - Navigating into Season 1 correctly shows 8 episodes watched but a 9th _unwatched_ episode.
- Selecting `Up Next` also correctly displays the 9th unwatched episode.

I have done the following:

- Cleared TMDB and Trakt caches
- Asserted in _both_ TMDB & Trakt APIs that the data exists.
- I have `Get watched indicators and episode counts (slower)` **enabled**
- I have `Next episodes checks calendar (slower but more accurate)` **enabled**

For example, here's the response I get from the `Get show watched progress` endpoint in the trakt API for that series:

```json
{
  "aired": 9,
  "completed": 8,
  "last_watched_at": "2023-11-21T00:57:56.000Z",
  "reset_at": {},
  "seasons": [
    {
      "number": 1,
      "title": {},
      "aired": 9,
      "completed": 8,
      "episodes": [
        {
          "number": 1,
          "completed": true,
          "last_watched_at": "2023-10-05T22:22:36.000Z"
        },
        {
          "number": 2,
          "completed": true,
          "last_watched_at": "2023-10-06T18:11:52.000Z"
        },
        {
          "number": 3,
          "completed": true,
          "last_watched_at": "2023-10-15T21:43:06.000Z"
        },
        {
          "number": 4,
          "completed": true,
          "last_watched_at": "2023-10-24T21:04:32.000Z"
        },
        {
          "number": 5,
          "completed": true,
          "last_watched_at": "2023-11-02T21:40:07.000Z"
        },
        {
          "number": 6,
          "completed": true,
          "last_watched_at": "2023-11-03T23:44:47.000Z"
        },
        {
          "number": 7,
          "completed": true,
          "last_watched_at": "2023-11-17T00:31:53.000Z"
        },
        {
          "number": 8,
          "completed": true,
          "last_watched_at": "2023-11-21T00:57:56.000Z"
        },
        {
          "number": 9,
          "completed": false,
          "last_watched_at": {}
        }
      ]
    }
  ],
  "hidden_seasons": [],
  "next_episode": {
    "season": 1,
    "number": 9,
    "title": "Week 8",
    "ids": {
      "trakt": 11031571,
      "tvdb": 10122393,
      "imdb": "tt29721388",
      "tmdb": 4890804,
      "tvrage": {}
    }
  },
  "last_episode": {
    "season": 1,
    "number": 8,
    "title": "Week 7",
    "ids": {
      "trakt": 11031570,
      "tvdb": 10112391,
      "imdb": "tt29703811",
      "tmdb": 4890800,
      "tvrage": {}
    }
  }
}
```

This is just one show and one episode, but it could be causing issues elsewhere as this may be an edgecase of some kind.

**Comment by jurialmunkey** [comment]
@florence-robert - Episode counts on the indicators are a separate matter. These are calculated entirely separately from the list itself so they won't impact what episodes display.

The reason for the difference is that they use (from memory) the extended watched history sync and sometimes it lags a little on the trakt sever cache refreshing with new info. That's why youre still seeing the wrong count even after deleting all local cache - because it is what the trakt server is saying. It's been a while since I wrote that code but I do distinctly remember there was good reason for doing it this way rather than using the in-progress end point per show (I can't remember the specifics but it was some combo of performance reasons and inprogress endpoint not giving all the data I needed)

Btw combined landscape - if you use this view type you should disable the kodi setting for parent folder items (..) as it doesn't work correctly with them in general (and nothing skin or plugin can do about it - it's just the nature of how that type of combined view needs to work).

**Comment by Garageflower82** [comment]
I'm having this issue. Was a solution found?

I have selected the option to check calendar. The calendar does show new episodes, but the "my next episodes" list does not

**Comment by HairyHook** [comment]
> I'm having this issue. Was a solution found?
> 
> I have selected the option to check calendar. The calendar does show new episodes, but the "my next episodes" list does not

In my case, it did not get fixed. The cache has to expire or you have to manually refresh it by watching something else and forcing the update that way.

**Comment by jurialmunkey** [comment]
> In my case, it did not get fixed. The cache has to expire or you have to manually refresh it by watching something else and forcing the update that way.

This should be addressed in the most recent version (6.7.7) with a shorter 12hr expiry (rather than 24hr).

I realised that the issue was that even with a 24hr expiry, the issue was that if you stopped watching at 11:59pm, then the 24hr expiry would mean it didn't expire until 11:59pm (since we last synced at 11:59). Obviously an extreme example, but point is that if you have a normal watching pattern of e.g. 7pm to 10pm then next sync isn't going to be until 10pm next day, which is after you normally start watching tv.

With a 12hr expiry its still long enough that we don't constantly resync needlessly but also short enough that we're checking *before* the next time we likely start watching (e.g. if you stop watching at 10pm, next resync is going to be 10am).

**Comment by Garageflower82** [comment]
> > In my case, it did not get fixed. The cache has to expire or you have to manually refresh it by watching something else and forcing the update that way.
> 
> This should be addressed in the most recent version (6.7.7) with a shorter 12hr expiry (rather than 24hr).
> 
> I realised that the issue was that even with a 24hr expiry, the issue was that if you stopped watching at 11:59pm, then the 24hr expiry would mean it didn't expire until 11:59pm (since we last synced at 11:59). Obviously an extreme example, but point is that if you have a normal watching pattern of e.g. 7pm to 10pm then next sync isn't going to be until 10pm next day, which is after you normally start watching tv.
> 
> With a 12hr expiry its still long enough that we don't constantly resync needlessly but also short enough that we're checking _before_ the next time we likely start watching (e.g. if you stop watching at 10pm, next resync is going to be 10am).

ok mate, thanks for the update. I'll change my widgets back to tmdb helper and hopefully all good. I appreciate your hard work

**Comment by HairyHook** [comment]
> > In my case, it did not get fixed. The cache has to expire or you have to manually refresh it by watching something else and forcing the update that way.
> 
> This should be addressed in the most recent version (6.7.7) with a shorter 12hr expiry (rather than 24hr).
> 
> I realised that the issue was that even with a 24hr expiry, the issue was that if you stopped watching at 11:59pm, then the 24hr expiry would mean it didn't expire until 11:59pm (since we last synced at 11:59). Obviously an extreme example, but point is that if you have a normal watching pattern of e.g. 7pm to 10pm then next sync isn't going to be until 10pm next day, which is after you normally start watching tv.
> 
> With a 12hr expiry its still long enough that we don't constantly resync needlessly but also short enough that we're checking _before_ the next time we likely start watching (e.g. if you stop watching at 10pm, next resync is going to be 10am).

Thank you so much for the explanation and the fix. 12h expiry is the perfect balance and should get rid of this issue permanently with my watching habits.

One quick thing, I was not running the latest (6.7.3 I think) but the moment that I updated to 6.7.7 and my tmdbhelper "trakt in progress shows" refreshed, it suddenly included all the shows that have been manually hidden in trakt. Any way to remove the hidden ones and only show the current trakt in progress list? Is this a new bug? Hidden shows were not visible previously.

**Comment by jurialmunkey** [comment]
> Thank you so much for the explanation and the fix. 12h expiry is the perfect balance and should get rid of this issue permanently with my watching habits.

Hopefully it is working correctly. I'm fairly sure it updated correctly for me last night in my loungeroom setup before watching something, but it is a bit difficult to test in a real world scenario as there's really only one chance per day for it! (and then only if I've actually got something new in my calendar and haven't caused a resync for a different reason).

> Any way to remove the hidden ones and only show the current trakt in progress list? Is this a new bug? Hidden shows were not visible previously.

Trakt is depreciating the hide feature for tvshows and replacing it with drop
https://forums.trakt.tv/t/goodbye-hide-hello-drop/51853

Hide still exists during the transition but I expect it to be removed entirely in the future for tvshows.

If you go to your hide settings in your profile, you should see a button to convert your hidden shows to be dropped with one click
https://trakt.tv/settings/hidden/watched/title

**Comment by HairyHook** [comment]
> > Thank you so much for the explanation and the fix. 12h expiry is the perfect balance and should get rid of this issue permanently with my watching habits.
> 
> Hopefully it is working correctly. I'm fairly sure it updated correctly for me last night in my loungeroom setup before watching something, but it is a bit difficult to test in a real world scenario as there's really only one chance per day for it! (and then only if I've actually got something new in my calendar and haven't caused a resync for a different reason).
> 
> > Any way to remove the hidden ones and only show the current trakt in progress list? Is this a new bug? Hidden shows were not visible previously.
> 
> Trakt is depreciating the hide feature for tvshows and replacing it with drop https://forums.trakt.tv/t/goodbye-hide-hello-drop/51853
> 
> Hide still exists during the transition but I expect it to be removed entirely in the future for tvshows.
> 
> If you go to your hide settings in your profile, you should see a button to convert your hidden shows to be dropped with one click https://trakt.tv/settings/hidden/watched/title

It is working great! As always, really appreciate your hard work and dedication. And thanks for mentioning the drop feature, I had no idea!
